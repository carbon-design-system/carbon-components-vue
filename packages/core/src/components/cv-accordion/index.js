import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
