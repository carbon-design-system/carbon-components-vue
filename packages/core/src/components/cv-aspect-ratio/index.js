import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
