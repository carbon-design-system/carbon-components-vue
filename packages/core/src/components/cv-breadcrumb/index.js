import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
