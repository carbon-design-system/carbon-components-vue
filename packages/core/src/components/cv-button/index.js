import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
