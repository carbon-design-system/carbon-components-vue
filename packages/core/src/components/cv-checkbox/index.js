import CvCheckbox from './cv-checkbox';
import CvCheckboxSkeleton from './cv-checkbox-skeleton';
export { CvCheckbox, CvCheckboxSkeleton };
