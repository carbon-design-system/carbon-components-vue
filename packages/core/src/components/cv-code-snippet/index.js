import CvCodeSnippet from './cv-code-snippet';
import CvCodeSnippetSkeleton from './cv-code-snippet-skeleton';
export { CvCodeSnippet, CvCodeSnippetSkeleton };
