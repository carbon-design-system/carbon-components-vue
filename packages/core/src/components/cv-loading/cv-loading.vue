<template>
  <cv-wrapper :tag-type="overlay ? 'div' : ''" class="cv-loading" :class="overlayClasses">
    <div
      data-loading
      :class="{
        'cv-loading': !overlay,
        [`${carbonPrefix}--loading`]: active || stopping,
        [`${carbonPrefix}--loading--stop`]: !active && stopping,
        [`${carbonPrefix}--loading--small`]: small,
      }"
      ref="loading"
    >
      <svg :class="`${carbonPrefix}--loading__svg`" viewBox="-75 -75 150 150">
        <title>Loading</title>
        <circle v-if="small" :class="`${carbonPrefix}--loading__background`" cx="0" cy="0" :r="loadingRadius" />
        <circle :class="`${carbonPrefix}--loading__stroke`" cx="0" cy="0" :r="loadingRadius" />
      </svg>
    </div>
  </cv-wrapper>
</template>

<script>
import CvWrapper from '../cv-wrapper/_cv-wrapper';
import { carbonPrefixMixin } from '../../mixins';

export default {
  name: 'CvLoading',
  mixins: [carbonPrefixMixin],
  components: { CvWrapper },
  props: {
    active: { type: Boolean, default: true },
    overlay: Boolean,
    small: Boolean,
  },
  computed: {
    overlayClasses() {
      const classes = [];
      if (this.overlay) {
        if (this.active || this.stopping) {
          classes.push(`${this.carbonPrefix}--loading-overlay`);
        } else {
          classes.push(`${this.carbonPrefix}--loading-overlay--stop`);
        }
      }

      return classes;
    },
    loadingRadius() {
      return this.small ? '26.8125' : '37.5';
    },
  },
  data() {
    return {
      stopping: false,
    };
  },
  watch: {
    active(val) {
      this.onActiveUpdate(val);
    },
  },
  methods: {
    onEnd(ev) {
      if (ev.animationName === 'rotate-end-p2') {
        this.$refs.loading.removeEventListener('animationend', this.onEnd);

        this.stopping = false;
        this.$emit('loading-end');
      }
    },
    onActiveUpdate(newValue) {
      this.stopping = !newValue;
      if (!newValue) {
        this.$refs.loading.addEventListener('animationend', this.onEnd);
      }
    },
  },
};
</script>
