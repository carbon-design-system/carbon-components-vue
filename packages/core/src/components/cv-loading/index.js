import CvLoading from './cv-loading';

export { CvLoading };
