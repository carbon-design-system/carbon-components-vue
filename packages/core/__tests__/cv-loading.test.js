// import { shallowMount as shallow, mount } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, mount, trigger, setProps } = awaitNextTick;
import { CvLoading } from '@/components/cv-loading';

describe('CvLoading', () => {
  const RADIUS_SMALL = '42';
  const RADIUS_NORMAL = '44';

  const loadingOverlayClass = 'bx--loading-overlay';
  const loadingOverlayStopClass = 'bx--loading-overlay--stop';
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvLoading, ['active', 'overlay', 'small'], Boolean);
  testComponent.propsHaveDefault(CvLoading, ['active']);

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('should render correctly when not active, with normal size and not overlayed', async () => {
    const propsData = { active: false, small: false, overlay: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when active, with normal size and not overlayed', async () => {
    const propsData = { active: true, small: false, overlay: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when not active, small and not overlayed', async () => {
    const propsData = { active: false, small: true, overlay: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when active, small and not overlayed', async () => {
    const propsData = { active: true, small: true, overlay: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when not active, with normal size and overlayed', async () => {
    const propsData = { active: false, small: false, overlay: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when active, with normal size and overlayed', async () => {
    const propsData = { active: true, small: false, overlay: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when not active, small and overlayed', async () => {
    const propsData = { active: false, small: true, overlay: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when active, small and overlayed', async () => {
    const propsData = { active: true, small: true, overlay: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  // ***************
  // FUNCTIONAL TESTS
  // ***************
  it('should compute loading radius correctly', async () => {
    const propsData = { small: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.vm.loadingRadius).toEqual(RADIUS_NORMAL);
  });

  it('should compute small loading radius correctly', async () => {
    const propsData = { small: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.vm.loadingRadius).toEqual(RADIUS_SMALL);
  });

  it('should compute `overlayClasses` correctly when overlay is not used', async () => {
    const propsData = { overlay: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.vm.overlayClasses).toEqual([]);
  });

  it('should compute `overlayClasses` correctly when overlayed and active', async () => {
    const propsData = { overlay: true, active: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.vm.overlayClasses).toEqual([loadingOverlayClass]);
  });

  it('should compute `overlayClasses` correctly when overlayed and stopping', async () => {
    const propsData = { overlay: true, active: true };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    expect(wrapper.vm.overlayClasses).toEqual([loadingOverlayClass]);
  });

  it('should compute `overlayClasses` correctly when overlayed and not active or stopped', async () => {
    const propsData = { overlay: true, active: false };
    const wrapper = await shallow(CvLoading, {
      propsData,
    });
    wrapper.setData({
      stopping: false,
    });
    expect(wrapper.vm.overlayClasses).toEqual([loadingOverlayStopClass]);
  });

  it('animation should be stopped once `active` prop is set to false', async () => {
    const propsData = { active: true };
    const wrapper = await mount(CvLoading, {
      propsData,
    });
    expect(wrapper.vm.stopping).toEqual(false);

    await setProps(wrapper, {
      active: false,
    });
    expect(wrapper.vm.stopping).toEqual(true);
    await trigger(wrapper, 'animationend', { animationName: 'rotate-end-p2' });
    expect(wrapper.vm.stopping).toEqual(false);
  });

  it('animation should be continued once `active` prop is set to true again', async () => {
    const propsData = { active: false };
    const wrapper = await mount(CvLoading, {
      propsData,
    });

    await setProps(wrapper, {
      active: true,
    });
    expect(wrapper.vm.stopping).toEqual(false);
  });

  it('should emit `loading-end` when animation `rotate-end-p2` stops', async () => {
    const propsData = { active: true };
    const wrapper = await mount(CvLoading, {
      propsData,
    });
    await setProps(wrapper, {
      active: false,
    });
    await trigger(wrapper, 'animationend', { animationName: 'rotate-end-p2' });
    expect(wrapper.emitted()['loading-end']).toBeTruthy();
  });

  it('should not emit `loading-end` when animation with the name differed from `rotate-end-p2` stops', async () => {
    const propsData = { active: true };
    const wrapper = await mount(CvLoading, {
      propsData,
    });
    await setProps(wrapper, {
      active: false,
    });
    await trigger(wrapper, 'animationend', { animationName: 'not-rotate-end-p2' });
    expect(wrapper.emitted()['loading-end']).toBeFalsy();
  });
});
