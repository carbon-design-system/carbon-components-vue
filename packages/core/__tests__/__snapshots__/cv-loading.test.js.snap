// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvLoading should render correctly when active, small and not overlayed 1`] = `
<cv-wrapper-stub tagtype="" class="cv-loading">
  <div data-loading="" class="cv-loading bx--loading bx--loading--small"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__background"></circle>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when active, small and overlayed 1`] = `
<cv-wrapper-stub tagtype="div" class="cv-loading bx--loading-overlay">
  <div data-loading="" class="bx--loading bx--loading--small"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__background"></circle>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when active, with normal size and not overlayed 1`] = `
<cv-wrapper-stub tagtype="" class="cv-loading">
  <div data-loading="" class="cv-loading bx--loading"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <!---->
      <circle cx="50%" cy="50%" r="44" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when active, with normal size and overlayed 1`] = `
<cv-wrapper-stub tagtype="div" class="cv-loading bx--loading-overlay">
  <div data-loading="" class="bx--loading"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <!---->
      <circle cx="50%" cy="50%" r="44" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when not active, small and not overlayed 1`] = `
<cv-wrapper-stub tagtype="" class="cv-loading">
  <div data-loading="" class="cv-loading bx--loading--small"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__background"></circle>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when not active, small and overlayed 1`] = `
<cv-wrapper-stub tagtype="div" class="cv-loading bx--loading-overlay--stop">
  <div data-loading="" class="bx--loading--small"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__background"></circle>
      <circle cx="50%" cy="50%" r="42" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when not active, with normal size and not overlayed 1`] = `
<cv-wrapper-stub tagtype="" class="cv-loading">
  <div data-loading="" class="cv-loading"><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <!---->
      <circle cx="50%" cy="50%" r="44" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;

exports[`CvLoading should render correctly when not active, with normal size and overlayed 1`] = `
<cv-wrapper-stub tagtype="div" class="cv-loading bx--loading-overlay--stop">
  <div data-loading="" class=""><label class="bx--visually-hidden">
      Loading
    </label> <svg viewBox="0 0 100 100" class="bx--loading__svg">
      <title>Loading</title>
      <!---->
      <circle cx="50%" cy="50%" r="44" class="bx--loading__stroke"></circle>
    </svg></div>
</cv-wrapper-stub>
`;
